provider/aws: Add support for AWS IoT

jhedev opened the pull request:
This is my very first PR to a go project so please be patient with me :)

This will add the following resources:
- `aws_iot_certificate`
- `aws_iot_thing`
- `aws_iot_policy`
- `aws_iot_policy_attachment`
- `aws_iot_topic_rule`

I already added acceptance tests for all resources except `aws_iot_policy_attachment` (I have a test locally but it is still failing...). 

Still todo:
- [x] Update function of `aws_iot_policy_attachment`
- [x] Update function of `aws_iot_topic_rule`
- [x] More tests for most resources
- [x] Clean up code
- [x] Documentation

Also see #6138

jhedev commented:
The `aws_iot_certificate` acceptance tests require a `csr.pem` file. This is still missing in the repo as I didn't know where to put it. Maybe someone can help me?

jhedev commented:
Added some basic documentation for all added resources.

jhedev commented:
Added the `csr.pem` file to the `test-fixtures` directory.

jhedev commented:
Rebased onto master. Unfortunately this messed up the commits a bit.

protomouse commented:
Thanks for putting work into this @jhedev. I've rebased [our fork](https://github.com/wowgroup/terraform/commits/feature/iot) against current master, and made some other (minor) changes to get the tests running again. Feel free to pull that into yours.

As for `aws_iot_policy_attachment`, perhaps it'd be less surprising if it follows the interface of [aws_iam_policy_attachment](https://www.terraform.io/docs/providers/aws/r/iam_policy_attachment.html)? That is: rather than attaching multiple policies to one principal, it attaches one policy to multiple principals. I'm not exactly sure why I had it the other way around, but I think it's more convenient also. You'd usually create one policy for a group of principals, so having the ability to perform this mapping in one go:

```hcl
resource "aws_iot_policy_attachment" {
  name = "pubsub-bulbs"
  policy = "${aws_iot_policy.pubsub.name}"
  principals = ["${aws_iot_certificate.bulb.*.arn}"]
}
```

seems easier to me.

jwestboston commented:
Looking forward to this being supported within Terraform!

jhedev commented:
Thanks @protomouse! I pulled your changes in. 

I remember wondering about your proposed interface of the `aws_iot_policy_attachment`, too. Not sure why I didn't ask, though. 

I also agree that this should be kept similar to the `aws_iam_policy_attachment`. So I'll update the code.

jhedev commented:
Rebased onto master and updated the interface of `aws_iot_policy_attachment` as discussed above. 

If there are no other remarks this can be merged now :)

jhedev commented:
Anything else I can do to speed up this PR getting merged?

protomouse commented:
Pinging @catsby for review.

jhedev commented:
Any chance you have some time to review, @catsby ? :)

protomouse commented:
I've resolved the merge conflicts in our fork, as well as brought the IoT dependency level with the other vendor'd AWS packages. Could you please @jhedev pull those changes into your branch, and then ping some more HashiCorp folk? (phinze, stack72, mitchellh)

jhedev commented:
@protomouse Done. Thanks for your help!

Any chance one of you can review, @catsby @phinze @stack72 @mitchellh ?

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+	d.SetId(*out.CertificateId)
+	d.Set("arn", *out.CertificateArn)
+
+	return nil
```
The normal convention here is for the Create -> Read func

This means we can check that the Create has happened correctly. All of the `d.Set("")` then happens in the Read func

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+	log.Printf("[DEBUG] Created certificate from csr")
+
+	d.SetId(*out.CertificateId)
+	d.Set("arn", *out.CertificateArn)
```
This can be removed as we set it in the Read func - we just need to make Create return the Read func

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+		return err
+	}
+
+	d.SetId(*out.CertificateDescription.CertificateId)
```
No need to reset the ID here

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+	}
+
+	d.SetId(*out.CertificateDescription.CertificateId)
+	d.Set("arn", *out.CertificateDescription.CertificateArn)
```
No need to deref, d.Set does this in a safe way

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+		}
+
+		//TODO: Remove old cert??
+		//conn.DeleteCertificate(&iot.DeleteCertificateInput{
```
IMO we should clean the old cert up *only* if we have actually created the certificate in the first place

Either way, let's clean up these comments

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+
+	if d.HasChange("csr") {
+		//Make old cert inactive
+		_, err := conn.UpdateCertificate(&iot.UpdateCertificateInput{
```
Should the order of operations change here? Should we create the new CSR first and then invalidate the old one? This would mean we don't Update a certificate until we know the new CSR has been successful

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+			return err
+		}
+
+		d.SetId(*out.CertificateId)
```
We don't need to set these here

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+		d.SetId(*out.CertificateId)
+		d.Set("arn", *out.CertificateArn)
+
+	} else {
```
If there are no changes to CSR, then we should specifically check for if `d.HasChange("active")` 

Otherwise we can potentially be making an unwanted change

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+		}
+	}
+
+	return nil
```
Update func usually calls Read func

stack72 commented on builtin/providers/aws/resource_aws_iot_certificate.go:
```diff
@@ -0,0 +1,152 @@
+
+	conn := meta.(*AWSClient).iotconn
+
+	_, err := conn.UpdateCertificate(&iot.UpdateCertificateInput{
```
Does this action take a period of time? If we try and do this and AWS is performing an action, then the Delete may fail

stack72 commented on builtin/providers/aws/resource_aws_iot_policy.go:
```diff
@@ -0,0 +1,139 @@
+	d.Set("arn", *out.PolicyArn)
+	d.Set("defaultVersionId", *out.PolicyVersionId)
+
+	return nil
```
Create -> read

stack72 commented on builtin/providers/aws/resource_aws_iot_policy.go:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	d.SetId(*out.PolicyName)
+	d.Set("arn", *out.PolicyArn)
```
Move d.Set to Read only

stack72 commented on builtin/providers/aws/resource_aws_iot_policy.go:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	d.SetId(*out.PolicyName)
+	d.Set("arn", *out.PolicyArn)
```
No need to deref

stack72 commented on builtin/providers/aws/resource_aws_iot_policy.go:
```diff
@@ -0,0 +1,139 @@
+			return err
+		}
+
+		d.Set("arn", *out.PolicyArn)
```
No need to reset these

stack72 commented on builtin/providers/aws/resource_aws_iot_policy.go:
```diff
@@ -0,0 +1,139 @@
+		d.Set("defaultVersionId", *out.PolicyVersionId)
+	}
+
+	return nil
```
Update -> Read

stack72 commented on builtin/providers/aws/resource_aws_iot_policy.go:
```diff
@@ -0,0 +1,139 @@
+func resourceAwsIotPolicyUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	//TODO: prune old versions
```
This should only happen on delete so we can remove this comment

stack72 commented on builtin/providers/aws/resource_aws_iot_policy_attachment.go:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	d.SetId(d.Get("name").(string))
+	d.Set("policy", d.Get("policy").(string))
```
Move d.Set to Read func

stack72 commented on builtin/providers/aws/resource_aws_iot_policy_attachment.go:
```diff
@@ -0,0 +1,151 @@
+	d.SetId(d.Get("name").(string))
+	d.Set("policy", d.Get("policy").(string))
+	d.Set("principals", d.Get("principals").(*schema.Set).List())
+	return nil
```
Create -> read

stack72 commented on builtin/providers/aws/resource_aws_iot_policy_attachment.go:
```diff
@@ -0,0 +1,151 @@
+	return &schema.Resource{
+		Create: resourceAwsIotPolicyAttachmentCreate,
+		Read:   resourceAwsIotPolicyAttachmentRead,
+		Update: resourceAwsIotPolicyAttachmentUpdate,
```
As this is an attachment, is there *really* an update? Wouldn't ForceNew (destroy attachment and then create a new attachment) be better?

stack72 commented on builtin/providers/aws/resource_aws_iot_policy_attachment.go:
```diff
@@ -0,0 +1,151 @@
+	conn := meta.(*AWSClient).iotconn
+
+	for _, p := range d.Get("principals").(*schema.Set).List() {
+		_, err := conn.AttachPrincipalPolicy(&iot.AttachPrincipalPolicyInput{
```
Is there a potential eventual consistency issue here? Do we need to wait and check the attachment has happened before we move on to the next?

stack72 commented on builtin/providers/aws/resource_aws_iot_policy_attachment.go:
```diff
@@ -0,0 +1,151 @@
+	return nil
+}
+
+func resourceAwsIotPolicyAttachmentUpdate(d *schema.ResourceData, meta interface{}) error {
```
I think we can make everything in the schema ForceNew and slip this

stack72 commented on builtin/providers/aws/resource_aws_iot_policy_attachment.go:
```diff
@@ -0,0 +1,151 @@
+
+	for _, p := range d.Get("principals").(*schema.Set).List() {
+		log.Printf("[INFO] %+v", p)
+		_, err := conn.DetachPrincipalPolicy(&iot.DetachPrincipalPolicyInput{
```
Is there a potential time to wait for the detachment to happen?

stack72 commented on builtin/providers/aws/resource_aws_iot_policy_attachment_test.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+func testAccCheckAWSIoTPolicyAttachmentDestroy_basic(s *terraform.State) error {
+	return nil
```
why have we got a func that returns nil?

jhedev replied:
The problem here is that I'm not sure how to test if the attachment was destroyed and just copied this idea from the `resource_aws_iam_policy_attachment_test.go` which does the same:

```go
func testAccCheckAWSPolicyAttachmentDestroy(s *terraform.State) error {
	return nil
}
```

fnouama commented:
👍 Looking forward to have this feature available through terraform

KoenR3 commented:
It is one of the downsides we currently have with terraform that we cannot provision IoT.  I hope this feature will be available soon!

jhedev commented:
@KoenR3 Same goes for me. Unfortunately I was pretty short on time the last couple months and couldn't make the requested changes. I hope to find some time in the next two weeks to make the required changes. So we can get this merged eventually :)

protomouse commented:
Cool @jhedev. I'm working on covering some of the more recent IoT functionality not covered by this PR, such as [thing types](https://aws.amazon.com/about-aws/whats-new/2016/07/aws-iot-announces-support-for-thing-types-in-the-thing-registry/) and [BYOC](https://aws.amazon.com/blogs/mobile/use-your-own-certificate-with-aws-iot/). I'll submit a separate PR for these. For now though, I'd like to suggest a small API change to `aws_iot_certificate` - namely, replacing the boolean attribute `active` with an enum `status` (`ACTIVE`, `INACTIVE`). This is because a certificate can also (irreversibly) be marked as `REVOKED`. Feel free to do as you wish w.r.t. actually implementing certificate revocation in this PR, but this API change will at least make adding this functionality in the future a bit smoother.

alessandroben commented:
Hi guys, do you have any news about that? I'm looking to implement aws_iot on my terraform project and I'm stuck since there is no support at the moment :( 
Let me know about the ETA of this feature, thanks you very much.

ajlanghorn commented:
IoT resources in Terraform would be awesome! Kudos for the great work on this so far.
Let me know if you need any help; happy to contribute, if I can!

theherk commented:
@jhedev, you are doing a great thing for many of us. If there were anything I could do to help I would, but you're in the driver's seat. How far do you believe you are from this being ready for merge. I want very much not to write a CloudFormation template for our IoT bits.

stephencoe commented on builtin/providers/aws/resource_aws_iot_policy_attachment.go:
```diff
@@ -0,0 +1,98 @@
+	for _, p := range d.Get("principals").(*schema.Set).List() {
+		_, err := conn.AttachPrincipalPolicy(&iot.AttachPrincipalPolicyInput{
+			Principal:  aws.String(d.Get("policy").(string)),
+			PolicyName: aws.String(p.(string)),
```
these variables are the wrong way around, they should be 
```
Principal:  aws.String(p.(string)),
PolicyName: aws.String(d.Get("policy").(string)),
```

stephencoe commented:
@jhedev thanks for the great work on this so far! I have been doing some testing and running your branch to build IOT within AWS. Everything has been great up to this point, I found some small inconsistencies with the dynamodb topic rule 

Missing fields `hash_key_type` & `range_key_type`. These are specified in the resource doc but were missed in the go file.

The [payload field](https://github.com/jhedev/terraform/blob/31417ffae64fc8363c0e743dc74cba6e4da2c1e8/builtin/providers/aws/resource_aws_iot_topic_rule.go#L109) shouldn't be `Required`. In the console / cli you are able to create this without applying the field. Having it required and setting it to an empty string causes the rule to error when writing the the stream (this part is a strange behaviour in AWS)

example iot_rule.json:
```json
{
    "sql": "<QUERY>",
    "ruleDisabled": false,
    "actions": [{
        "dynamoDB": {
            "tableName": "<TABLE>",
            "roleArn": "<ROLE ARN>",
            "hashKeyField": "unique_id",
            "hashKeyValue": "${id}",
            "rangeKeyField": "timestamp",
            "rangeKeyValue": "${timestamp()}"
        }
    }]
}
```
`aws iot create-topic-rule --rule-name dynamo_rule_test --topic-rule-payload file://iot_rule.json`

Let us know if there is anything we can do to help

jhedev commented:
@stephencoe Thanks for testing! That helps a lot.

Sorry for being so quiet lately. I didn't have much time to work on this but hope to find some time to  work on it during the following weekend.

Reviewing code and testing is always very much appreciated :)

jhedev commented:
I made some of the changes mentioned above and rebased on master again. However, there are still a few things that need to be change but I'm not sure how. So it would be great to get some input how these things are handled in terraform and possible a pointer to some source code where this is done:

* `aws_iot_policy_attachment`:
    * [ ] Eventual consistency issue when attaching principals? see [discussion](https://github.com/hashicorp/terraform/pull/6961#discussion_r90891886)
    * [ ] Wait some time when detaching? see [discussion](https://github.com/hashicorp/terraform/pull/6961#discussion_r90892115)

* `aws_iot_thing`:
    * [ ] Same issue as above when attaching principals?
    * [ ] Same issue as above when detaching principals?

stephencoe commented:
@jhedev this is awesome! I don't have a great deal of experience with the code base but looking at other examples where this has been implemented it looks like it's wrapped in a `resource.Retry`. 

Some examples I found of this were in [lambda_function](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_lambda_function.go#L307), [api_gateway_account](https://github.com/hashicorp/terraform/blob/351c6bed79abbb40e461d3f7d49fe4cf20bced41/builtin/providers/aws/resource_aws_api_gateway_account.go#L100), [aws_launch_configuration](https://github.com/hashicorp/terraform/blob/affce99a8ee70150eaf169e489e779843126a86c/builtin/providers/aws/resource_aws_launch_configuration.go).

jhedev commented:
@stephencoe Thanks for your help. I'll have a look at the implementations.

rob-smallshire commented:
Is this being worked on currently by anybody? No activity for a few months now...

jhedev commented:
Hey @rob-smallshire, I always planned to get this finished at some point. However, my free time is very rare at the moment and since I don't use AWS IoT at work anymore there is very little motivation to get this done.

Long story short: I do not work on it actively at the moment and don't think I will in the near future.

Feel free to take over and use any of my code if you like :)

rob-smallshire commented:
@jhedev Thanks for a prompt response. It's good to have some clarity on the status of this feature. I'm still evaluating AWS IoT for my application, but if it works out for us I can dedicate some effort towards resurrecting your code. It looks like your were pretty close!

jhedev commented:
I guess we should close this PR now?

rob-smallshire commented:
I guess so. Work continues over here:

https://github.com/terraform-providers/terraform-provider-aws/issues/143

and

https://github.com/terraform-providers/terraform-provider-aws/pull/986

jhedev commented:
Yeah, I've seen that. Thanks for taking over :)
